src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -0,0 +1,146 @@
+package com.pusher.client.channel.impl;
+
+import com.google.gson.Gson;
+import com.pusher.client.AuthorizationFailureException;
+import com.pusher.client.Authorizer;
+import com.pusher.client.channel.PrivateEncryptedChannel;
+import com.pusher.client.channel.PrivateEncryptedChannelEventListener;
+import com.pusher.client.channel.SubscriptionEventListener;
+import com.pusher.client.connection.impl.InternalConnection;
+import com.pusher.client.crypto.nacl.SecretBoxOpener;
+import com.pusher.client.util.Base64;
+import com.pusher.client.util.Factory;
+
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class PrivateEncryptedChannelImpl extends ChannelImpl implements PrivateEncryptedChannel {
+
+    private static final Gson GSON = new Gson();
+    private final InternalConnection connection;
+    private final Authorizer authorizer;
+    private SecretBoxOpener secretBoxOpener;
+
+    private class PrivateEncryptedChannelData {
+        byte[] auth;
+        final String channelData;
+
+        protected PrivateEncryptedChannelData(String auth, String channelData) {
+            this.auth = auth.getBytes();
+            this.channelData = channelData;
+        }
+
+        public String getAuth() {
+            return new String(auth);
+        }
+
+        public String getChannelData() {
+            return channelData;
+        }
+
+        protected void clearAuthToken(){
+            Arrays.fill(auth, (byte)0);
+        }
+    }
+
+    private PrivateEncryptedChannelData authorizerData;
+
+    public PrivateEncryptedChannelImpl(final InternalConnection connection,
+                                       final String channelName,
+                                       final Authorizer authorizer,
+                                       final Factory factory) {
+        super(channelName, factory);
+        this.connection = connection;
+        this.authorizer = authorizer;
+    }
+
+    /* PrivateChannel implementation */
+
+
+    /* Base class overrides */
+
+    @Override
+    public void bind(final String eventName, final SubscriptionEventListener listener) {
+
+        if (!(listener instanceof PrivateEncryptedChannelEventListener)) {
+            throw new IllegalArgumentException(
+                    "Only instances of PrivateEncryptedChannelEventListener can be bound " +
+                            "to a private encrypted channel");
+        }
+
+        super.bind(eventName, listener);
+    }
+
+    private void saveSharedSecret(String sharedSecret) {
+        secretBoxOpener = new SecretBoxOpener(Base64.decode(sharedSecret));
+
+        // todo can we clear everything when the user disconnects totally?
```

Comment:
this piece of work is already captured in the clear token ticket - should i keep the todo here? or are we okay to remove this in code?